app/utils/practice-system.ts

```diff
@@ -109,42 +111,13 @@ export class PracticeSystem {
     } = this.deck;
 
     if (randomMode) {
-      const result: PracticeEntryTable = await Q.practiceEntries()
-        .select("practiceEntries.*", {
-          // uniform random with "id" and "updatedAt" computed by
-          //   rand(hash(id) ^ hash(updatedAt) ^ seed)
-          __uniform__: client.raw(
-            "RAND(CAST(CONV(SUBSTRING(HEX(UNHEX(SHA1(id)) ^ UNHEX(SHA1(updatedAt)) ^ __subQuery__.__seed__), 1, 8), 16, 10) as UNSIGNED))"
-          ),
-        })
-        .where("practiceEntries.deckId", deckId)
-        .where("practiceEntries.scheduledAt", "<=", now)
-        .crossJoin(
-          client.raw(
-            // global seed `hash(hash(max(updatedAt)))` which satisfies a desired property:
-            //   a seed should be updated on each `createPracticeAction`
-            "(SELECT UNHEX(SHA1(SHA1(updatedAt))) as __seed__, RAND(CAST(CONV(SUBSTRING(SHA1(updatedAt), 1, 8), 16, 10) as UNSIGNED)) as __seed_uniform__ FROM practiceEntries where deckId = ? ORDER BY updatedAt DESC LIMIT 1) as __subQuery__",
-            deckId
-          )
-        )
-        .orderByRaw(
-          // tweak the distribution based on
-          // - queueType
-          // - scheduledAt
-          client.raw(
-            `
-            (
-              __uniform__
-              + (queueType = 'NEW'   ) * -10 * (__subQuery__.__seed_uniform__ <= 0.80)
-              + (queueType = 'LEARN' ) * -10 * (__subQuery__.__seed_uniform__ >  0.80) * (__subQuery__.__seed_uniform__ <= 0.95)
-              + (queueType = 'REVIEW') * -10 *                                           (__subQuery__.__seed_uniform__ >  0.95)
-              + LEAST(-0.5, 0.1 / (60 * 60 * 24 * 7) * (UNIX_TIMESTAMP(scheduledAt) - UNIX_TIMESTAMP(?)))
-            )`,
-            now
-          )
-        )
-        .first();
-      return result;
+      const { query } = queryNextPracticeEntryRandomMode(
+        this.deck.id,
+        now,
+        this.deck.updatedAt.getTime()
```

Comment:
Actually we cannot use `deck.updatedAt` as a seed since it's not always updated during `createPracticeAction`

https://github.com/hi-ogawa/ytsub-v3/blob/8b3153a0a54317d7af809ad83aa451c756e47df6/app/utils/practice-system.ts#L257-L264

Reply:
For now we can deal with by extra query https://github.com/hi-ogawa/ytsub-v3/pull/282/commits/bed8cea63e14c4356ee7b098651cf31eb603efca